### DISCUSSION TITLE
How to create Azure AD app using service principal credentials with Microsoft Graph API

### DISCUSSION THREAD

**Discussion by collin-catalogic** [microsoftgraph/msgraph-sdk-go#84] [Q&A] [answered]
I am trying to migrate from Azure AD to MSGraph, and the application is built in Golang.
Can someone please point me to an example/snippet/documentation where i can know how to create an AD application using msgraph-sdk-go package?
Thanks in advance.

**Comment by baywet** [comment, marked as answer]
Hi @collin-catalogic
Thanks for trying the Go SDK and for reaching out.
The Azure AD Identity POD for kuberntes does just that and you [can have a look at their code](https://github.com/Azure/azure-workload-identity/blob/2618ac4ac34232a965f2394d0b6538928530cae3/pkg/cloud/graph.go#L24)
In their case they are using the beta SDK because they need access to other operations that only exist in beta, but you can perfectly do that with the v1 (this) sdk, [see the documentation snippet](https://docs.microsoft.com/en-us/graph/api/serviceprincipal-post-serviceprincipals?view=graph-rest-1.0&tabs=go)

**Reply by collin-catalogic** [reply]
Thanks a lot!!

**Reply by collin-catalogic** [reply]
@baywet I am vetting out this SDK for v1, are the functions like create, list keys of Storage Account and creation of Containers supported by this SDK.
Thanks,

**Reply by baywet** [reply]
by storage accounts, do you mean azure blob storage? or something else?

**Reply by collin-catalogic** [reply]
Yes, storage account that contains the blobs, files etc.

**Reply by baywet** [reply]
Azure services can be access through the [Azure SDK for Go](https://github.com/Azure/azure-sdk-for-go/), this SDK doesn't not compete with the Microsoft Graph Go SDK (this repo) which gives you access to Microsoft 365 (AAD+Office 365+Intune+...), they are complementary.
The only APIs that might overlap are the identity/AAD ones, but I recommend you use the MS Graph SDK for those as more APIs are exposed (not just application registrations and service principals but also groups, roles, users and much more).

Hopefully this clarifies the use case for both SDKs.

**Reply by collin-catalogic** [reply]
Yes understood, but since Azure AD Graph would be soon deprecated, would using "Azure SDK for Go" for Azure services work?

**Reply by baywet** [reply]
I thought Azure SDK for Go uses ARM APIs as far as I understand, hence why they offer fewer operations/endpoints/entities. I believe these APIs were talking to AAD Graph under the covers but have been updated (are in the process) to talk to Microsoft Graph instead.

This code segment supports that assumption.

https://github.com/Azure/azure-sdk-for-go/blob/539bb880b057d30cbd83fc9e6003ae58f65d4f8d/services/aad/mgmt/2017-04-01/aad/client.go#L18

However [this post](https://github.com/Azure/azure-sdk-for-go/issues/7352) would infer the Azure Go SDK is actually directly talking to AAD Graph, which I couldn't find any support for in the code.

I think the safest at this point for you would be to use the Microsoft Graph Go SDK, this is not going anywhere anytime soon, and it offers more capabilities for the identity workload :)

**Reply by collin-catalogic** [reply]
@baywet Thank you for your time and help. 
Just one small query, I dont see [storage account](https://docs.microsoft.com/en-us/azure/storage/common/storage-account-overview) creation, deletion APIs in this repo, will those be part of the SDK?

**Reply by baywet** [reply]
No it won't, use the Azure SDK for azure workloads (except for directory services) and the Microsoft Graph SDK for Microsoft 365 workloads (Office 365, Intune, AAD,...)

**Reply by collin-catalogic** [reply]
Sure, thank you!!
